Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,588 @@
+        mult (OddPrime p k g n) = OddPrime p k g (s `stimes` n)
+        mult (TwoPower k a b) = TwoPower k (s `stimes` a) (s `stimes` b)
+        mult Two = Two
+
+-- | We define `succ` and `pred` with more efficient implementations than
+-- @`toEnum` . (+1) . `fromEnum`@.
+instance KnownNat n => Enum (DirichletCharacter n) where
+  toEnum = indexToChar
+  fromEnum = fromIntegral . characterNumber
+  succ x = makeChar x (characterNumber x + 1)
+  pred x = makeChar x (characterNumber x - 1)
+
+  enumFromTo x y       = bulkMakeChars x [fromEnum x..fromEnum y]
+  enumFrom x           = bulkMakeChars x [fromEnum x..]
+  enumFromThenTo x y z = bulkMakeChars x [fromEnum x, fromEnum y..fromEnum z]
+  enumFromThen x y     = bulkMakeChars x [fromEnum x, fromEnum y..]
+
+instance KnownNat n => Bounded (DirichletCharacter n) where
+  minBound = indexToChar (0 :: Int)
+  maxBound = indexToChar (totient n - 1)
+    where n = natVal (Proxy :: Proxy n)
+
+-- | We have a (non-canonical) enumeration of dirichlet characters.
+characterNumber :: DirichletCharacter n -> Integer
+characterNumber (Generated y) = foldl' go 0 y
+  where go x (OddPrime p k _ a) = x * m + numerator (fromRootOfUnity a * (m % 1))
+          where p' = fromIntegral (unPrime p)
+                m = p'^(k-1)*(p'-1)
+        go x (TwoPower k a b)   = x' * 2 + numerator (fromRootOfUnity a * (2 % 1))
+          where m = bit (k-2) :: Integer
+                x' = x `shiftL` (k-2) + numerator (fromRootOfUnity b * fromIntegral m)
+        go x Two = x
+
+-- | Give the dirichlet character from its number.
+-- Inverse of `characterNumber`.
+indexToChar :: forall n a. (KnownNat n, Integral a) => a -> DirichletCharacter n
+indexToChar = runIdentity . indicesToChars . Identity
+
+-- | Give a collection of dirichlet characters from their numbers. This may be more efficient than
+-- `indexToChar` for multiple characters, as it prevents some internal recalculations.
+indicesToChars :: forall n a f. (KnownNat n, Integral a, Functor f) => f a -> f (DirichletCharacter n)
+indicesToChars = fmap (Generated . unroll t . (`mod` m) . fromIntegral)
+  where n = natVal (Proxy :: Proxy n)
+        (Product m, t) = mkTemplate n
+
+-- | List all characters for the modulus. This is preferred to using @[minBound..maxBound]@.
+allChars :: forall n. KnownNat n => [DirichletCharacter n]
+allChars = indicesToChars [0..m-1]
+  where m = totient $ natVal (Proxy :: Proxy n)
+
+-- | The same as `indexToChar`, but if we're given a character we can create others more efficiently.
+makeChar :: Integral a => DirichletCharacter n -> a -> DirichletCharacter n
+makeChar x = runIdentity . bulkMakeChars x . Identity
+
+-- | Use one character to make many more: better than indicesToChars since it avoids recalculating
+-- some primitive roots
+bulkMakeChars :: (Integral a, Functor f) => DirichletCharacter n -> f a -> f (DirichletCharacter n)
+bulkMakeChars x = fmap (Generated . unroll t . (`mod` m) . fromIntegral)
```

Comment:
Again, just consume `f Natural` instead of `f a` and eliminate `fromIntegral`.

Reply:
This function is not exposed but is used internally in multiple contexts, so if it's okay I'll leave this one polymorphic.

Reply:
OK